blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+        "highlight.project_id": "<YOUR_PROJECT_ID>",
+        "deployment.environment.name": process.env.NODE_ENV,
+      },
+      propagators: ["auto"],
+      spanProcessors: [
+        new SimpleSpanProcessor(
+          new OTLPTraceExporter({
+            url: "https://otel.highlight.io:4317/v1/traces",
+            keepAlive: true,
+          }),
+        ),
+      ],
+      logRecordProcessor: new SimpleLogRecordProcessor(
+        new OTLPLogExporter({
+          url: "https://otel.highlight.io:4317/v1/logs",
+          keepAlive: true,
+        }),
+      ),
+      metricReader: new PeriodicExportingMetricReader({
+        exporter: new OTLPMetricExporter({
+          url: "https://otel.highlight.io:4317/v1/metrics",
+          keepAlive: true,
+        }),
+      }),
+    });
+  }
+}
+```
+
+Now, let's use the OpenTelemetry SDK in our route to emit data:
+
+```typescript
+import { NextResponse } from "next/server";
+import api, { propagation } from "@opentelemetry/api";
+import { logs, SeverityNumber } from "@opentelemetry/api-logs";
+
+export async function GET() {
+  const { email, name } = req.query;
+
+  // Here, you would typically make a call to your Python service
+  // passing the user's email or other identifier
+  // For now, we'll just log the information and return mock data
+  console.log(`Fetching data for user: ${email} ${name}`);
+
+  // In a real implementation, you would fetch data from the Garmin Health API here
+  const tracerProvider = api.trace.getTracerProvider();
+  const tracer = tracerProvider.getTracer("data");
+  const data = await tracer.startActiveSpan(
+    "data.fetch",
+    {
+      attributes: {
+        "user.email": email || undefined,
+        "user.name": name || undefined,
+      },
+    },
+    async () => {
+      console.log("Fetching data...", { email });
+      const headers = {
+        "Content-Type": "application/json",
+        Authorization: `Basic ${process.env.PYTHON_API_SECRET}`,
+      };
+      propagation.inject(api.context.active(), headers);
+      const response = await fetch(
+        `${process.env.PYTHON_API_HOSTNAME}/users/scrape`,
+        {
+          method: "POST",
+          headers,
+          body: JSON.stringify({
+            email,
+          }),
+        },
+      );
+      if (!response.ok) {
+        throw new Error("Failed to fetch data");
+      }
+      return await response.json();
+    },
+  );
+
+  const meterProvider = api.metrics.getMeterProvider();
+  const meter = meterProvider.getMeter("data");
+  const gauge = meter.createObservableGauge("data.metric");
+  for (const d of data) {
+    gauge.addCallback((m) => {
+      m.observe(d.attribute);
+    });
+  }
+
+  const provider = logs.getLoggerProvider();
+  const logger = provider.getLogger("data");
+  logger.emit({
+    severityNumber: SeverityNumber.INFO,
+    severityText: "INFO",
+    body: "returning data",
+    attributes: { data },
+  });
+
+  return NextResponse.json(data);
```

Comment:
Sorry, but this code example is confusing to me... Isn't most of what we're doing here already handled by `registerOtel`? What do you think about showing off some examples of the data collected and then add a couple additional details showing how to emit a log and metrics to enhance the data we're already getting?

Also, WDYT about showing a cleaner API for metrics and logs where we create a utility to provide a better DX to emit logs/metrics?

Reply:
good idea - cleaned this up to add an explanation and added an example using the `@highlight-run/node` sdk